src/pubsub/ua_pubsub_securitygroup.c

```diff
@@ -0,0 +1,322 @@
+                           UA_NodeId *securityGroupNodeId) {
+    if(!securityGroupConfig)
+        return UA_STATUSCODE_BADINVALIDARGUMENT;
+
+    /*check minimal config parameters*/
+    if(!securityGroupConfig->keyLifeTime ||
+       UA_String_isEmpty(&securityGroupConfig->securityGroupName) ||
+       UA_String_isEmpty(&securityGroupConfig->securityPolicyUri))
+        return UA_STATUSCODE_BADINVALIDARGUMENT;
+
+    if(UA_SecurityGroup_findSGbyName(server, securityGroupConfig->securityGroupName))
+        return UA_STATUSCODE_BADNODEIDEXISTS;
+
+    UA_PubSubSecurityPolicy *policy = NULL;
+    UA_Server_findPubSubSecurityPolicy(server, &securityGroupConfig->securityPolicyUri,
+                                       &policy);
+    if(!policy)
+        return UA_STATUSCODE_BADSECURITYPOLICYREJECTED;
+
+    if(securityGroupConfig->securityGroupName.length > 512)
+        return UA_STATUSCODE_BADOUTOFMEMORY;
+
+    UA_StatusCode retval = UA_STATUSCODE_BAD;
+
+    UA_SecurityGroup *newSecurityGroup =
+        (UA_SecurityGroup *)UA_calloc(1, sizeof(UA_SecurityGroup));
+    if(!newSecurityGroup)
+        return UA_STATUSCODE_BADOUTOFMEMORY;
+
+    memset(newSecurityGroup, 0, sizeof(UA_SecurityGroup));
+    UA_SecurityGroupConfig_copy(securityGroupConfig, &newSecurityGroup->config);
+
+    retval = UA_NodeId_copy(&securityGroupFolderNodeId,
+                            &newSecurityGroup->securityGroupFolderId);
+    if(retval != UA_STATUSCODE_GOOD) {
+        UA_SecurityGroup_delete(newSecurityGroup);
+        return retval;
+    }
+
+    retval = UA_String_copy(&securityGroupConfig->securityGroupName,
+                            &newSecurityGroup->securityGroupId);
+    if(retval != UA_STATUSCODE_GOOD) {
+        UA_SecurityGroup_delete(newSecurityGroup);
+        return retval;
+    }
+
+    retval = initializeKeyStorageWithKeys(server, newSecurityGroup);
+    if(retval != UA_STATUSCODE_GOOD) {
+        UA_SecurityGroup_delete(newSecurityGroup);
+        return retval;
+    }
+
+#ifdef UA_ENABLE_PUBSUB_INFORMATIONMODEL
+    retval = addSecurityGroupRepresentation(server, newSecurityGroup);
+    if(retval != UA_STATUSCODE_GOOD) {
+        UA_LOG_ERROR(&server->config.logger, UA_LOGCATEGORY_SERVER,
+                     "Add SecurityGroup failed with error: %s.",
+                     UA_StatusCode_name(retval));
+        UA_SecurityGroup_delete(newSecurityGroup);
+        return retval;
+    }
+#else
+    UA_PubSubManager_generateUniqueNodeId(&server->pubSubManager,
+                                          &newSecurityGroup->securityGroupNodeId);
+#endif
+
+    if(securityGroupNodeId)
+        UA_NodeId_copy(&newSecurityGroup->securityGroupNodeId, securityGroupNodeId);
+
+    if(server->pubSubManager.securityGroupsSize != 0) {
+        TAILQ_INSERT_TAIL(&server->pubSubManager.securityGroups, newSecurityGroup,
+                          listEntry);
+    } else {
+        TAILQ_INIT(&server->pubSubManager.securityGroups);
```

Comment:
You should call TAILQ_INIT in the _init of the PubSubManager.
Then you don't need to differentiate between cases here at all. Just do a INSERT_TAIL.

Reply:
I anyhow have to check both cases because if list is empty then TAILQ requires INSERT_HEAD and in rest of the cases INSERT_TAIL is used.
Also I checkout other places where TAILQ_INIT is used such as shown in the following. TAILQ is used in this manner. So I simply used as it was used for other cases like connections
https://github.com/open62541/open62541/blob/779e6b7902ff223b27d402c952fc5f06e967dd23/src/pubsub/ua_pubsub_manager.c#L78

also if calling TAILQ_INIT in the _init of the PubSubManager, would result in initializing securitygroups list every time whether the Security Groups will be used or not. Therefore I did the initialization when the first SecurityGroup is added to the server only then the list should be initialized.

Reply:
>  if list is empty then TAILQ requires INSERT_HEAD and in rest of the cases INSERT_TAIL is used.

What gives you that impression?
Reading the code in queue.h, INSERT_TAIL always works once the structure is correctly initialized.

Reply:
> > if list is empty then TAILQ requires INSERT_HEAD and in rest of the cases INSERT_TAIL is used.
> 
> What gives you that impression? Reading the code in queue.h, INSERT_TAIL always works once the structure is correctly initialized.

I tested it and now it works with INSERT_TAIL. However, at other places as mentioned above it is used as the same way I did. May be it should be improved at other places like `UA_PubSubManager_addConnection`, and  `UA_PubSubManager_addTopic`

Regarding INIT, It added to pubsub UA_PubSubManager_init `UA_PubSubManager_init` with `#UA_ENABLE_SKS`. and I think this can done for other PubSub related lists.?

Reply:
Ok. let's improve the other places then as well.
But not in this PR.